### PR TITLE
ProgressBarLogger UX Enhancements

### PR DESCRIPTION
**Pull request by ravi-mosaicml**
## Overview

* The ProgressBarLogger had a few bugs which caused progress bars to jump around on the terminal. This was because the `close` closed the training bar on eval end, and the position argument was set incorrectly for epoch-wise evaluation.
* The dataloader label is included as part of the progress bar output. This is helpful if using multiple evaluators.
* If evaluating mid-epoch, including the batch number as part of the progress bar label (see example below)
* Cleaned up the implementation to remove the `self.is_train` variable and `self._current_pbar`, since these likely lead to problems with the jumping.
* Use `dynamic_ncols` only if local, since k8s doesn't know the output terminal size. If on k8s, limit the max width to 120 characters. This ensures that the right-most remnants of a previous progress bar are not left over overriding position=1 with a new progress bar.
* Verified that progress bars display correctly both when running locally and when running over mcli (this only took 5 hours to find a combination of parameters that worked for both lol)

I didn't include test cases to verify the fix, as it's hard to verify visual improvements.

## Examples
The static examples below don't do full justice, since they show how the progress bars appear at the end. I'd recommend cloning this fork and running the yamls to verify that all the updating formats nicely.

### Max duration in epochs

YAML:
```yaml
train_dataset:
  &train_dataset
  mnist:
    use_synthetic: true
evaluators:
  eval1:
    label: eval/batch
    eval_interval: 1ba
    subset_num_batches: 2
    eval_dataset:
      <<: *train_dataset
  eval2:
    label: eval/ep
    eval_interval: 1ep
    subset_num_batches: 2
    eval_dataset:
      <<: *train_dataset
model:
  mnist_classifier:
    num_classes: 10
max_duration: 2ep
train_batch_size: 2000
eval_batch_size: 2000
dataloader:
  pin_memory: true
  timeout: 0
  prefetch_factor: 2
  persistent_workers: true
  num_workers: 1
train_subset_num_batches: 2
```

Output:
```
eval/batch     Batch   1:  100%|█████████████████████████| 2/2 [00:00<00:00,  4.65ba/s]          
eval/batch     Batch   2:  100%|█████████████████████████| 2/2 [00:00<00:00,  6.66ba/s]          
train          Epoch   0:  100%|█████████████████████████| 2/2 [00:01<00:00,  1.41ba/s, loss/trai
eval/ep        Epoch   0:  100%|█████████████████████████| 2/2 [00:00<00:00,  5.33ba/s]          
eval/batch     Batch   3:  100%|█████████████████████████| 2/2 [00:00<00:00,  5.74ba/s]          
eval/batch     Batch   4:  100%|█████████████████████████| 2/2 [00:00<00:00,  6.23ba/s]          
train          Epoch   1:  100%|█████████████████████████| 2/2 [00:01<00:00,  1.53ba/s, loss/trai
eval/ep        Epoch   1:  100%|█████████████████████████| 2/2 [00:00<00:00,  5.97ba/s] 
```

### Max duration in batches

YAML
```yaml
train_dataset:
  &train_dataset
  mnist:
    use_synthetic: true
evaluators:
  eval1:
    label: eval/batch
    eval_interval: 4ba
    subset_num_batches: 2
    eval_dataset:
      <<: *train_dataset
  eval2:
    label: eval/ep
    eval_interval: 1ep
    subset_num_batches: 2
    eval_dataset:
      <<: *train_dataset
model:
  mnist_classifier:
    num_classes: 10
max_duration: 20ba
train_batch_size: 2000
eval_batch_size: 2000
dataloader:
  pin_memory: true
  timeout: 0
  prefetch_factor: 2
  persistent_workers: true
  num_workers: 1
train_subset_num_batches: 10
```

Output:
```
eval/batch     Batch      4:  100%|█████████████████████████| 2/2 [00:00<00:00,  6.18ba/s]       
eval/batch     Batch      8:  100%|█████████████████████████| 2/2 [00:00<00:00,  6.23ba/s]       
eval/ep        Batch     10:  100%|█████████████████████████| 2/2 [00:00<00:00,  5.30ba/s]       
eval/batch     Batch     12:  100%|█████████████████████████| 2/2 [00:00<00:00,  6.16ba/s]       
eval/batch     Batch     16:  100%|█████████████████████████| 2/2 [00:00<00:00,  5.73ba/s]       
eval/batch     Batch     20:  100%|█████████████████████████| 2/2 [00:00<00:00,  5.53ba/s]       
eval/ep        Batch     20:  100%|█████████████████████████| 2/2 [00:00<00:00,  6.10ba/s]       
train                         100%|█████████████████████████| 20/20 [00:10<00:00,  2.08ba/s, loss
```

### Max duration in samples (and eval at fit end)

YAML:
```yaml
train_dataset:
  &train_dataset
  mnist:
    use_synthetic: true
evaluators:
  eval1:
    label: eval/batch
    eval_interval: 1ba
    subset_num_batches: 2
    eval_dataset:
      <<: *train_dataset
  eval2:
    label: eval/ep
    eval_interval: 1ep
    subset_num_batches: 2
    eval_dataset:
      <<: *train_dataset
model:
  mnist_classifier:
    num_classes: 10
max_duration: 6000sp
train_batch_size: 2000
eval_batch_size: 2000
dataloader:
  pin_memory: true
  timeout: 0
  prefetch_factor: 2
  persistent_workers: true
  num_workers: 1
train_subset_num_batches: 2
```

Output:
```
eval/batch     Sample  2000:  100%|█████████████████████████| 2/2 [00:00<00:00,  5.03ba/s]                                                                                                           
eval/batch     Sample  4000:  100%|█████████████████████████| 2/2 [00:00<00:00,  5.17ba/s]                                                                                                           
eval/ep        Sample  4000:  100%|█████████████████████████| 2/2 [00:00<00:00,  4.26ba/s]                                                                                                           
eval/batch     Sample  6000:  100%|█████████████████████████| 2/2 [00:00<00:00,  5.23ba/s]                                                                                                           
eval/ep        Sample  6000:  100%|█████████████████████████| 2/2 [00:00<00:00,  4.10ba/s]                                                                                                           
train                         100%|█████████████████████████| 6000/6000 [00:03<00:00, 1798.40sp/s, loss/train=2.2975]
```

Closes https://mosaicml.atlassian.net/browse/CO-633

### CODE REVIEW

File: composer/loggers/progress_bar_logger.py [hunk 1]
```diff
@@ -163,18 +168,9 @@ def __init__(
...
-
-    @_current_pbar.setter
-    def _current_pbar(self, pbar: Optional[_ProgressBar]):
-        assert self.is_train is not None, 'Cannot set pbar if self.is_train is not set.'
-
-        if self.is_train:
-            self.train_pbar = pbar
-        else:
-            self.eval_pbar = pbar
+        if self.eval_pbar is not None:
```
Comment (hanlint) [inline composer/loggers/progress_bar_logger.py:171]: if we remove `self.is_train`, then we should also just remove `_current_pbar` and have the code be explicit about which bar they are using.

File: composer/loggers/progress_bar_logger.py [hunk 2]
```diff
@@ -294,17 +295,36 @@ def eval_batch_end(self, state: State, logger: Logger) -> None:
...
     def epoch_end(self, state: State, logger: Logger) -> None:
         # only close the progress bar if its epoch_style
+        # Otherwise, the same progress bar is used for all of training, so do not close it here
         if self.train_pbar and self.train_pbar.epoch_style:
-            self._close()
+            self.train_pbar.close()
+            self.train_pbar = None
+
+    def fit_end(self, state: State, logger: Logger) -> None:
+        # If the train pbar isn't closed (i.e. not epoch style), then it would still be open here
```
Comment (hanlint) [inline composer/loggers/progress_bar_logger.py:304]: This comment could be simplified to: close all bars at fit end

**Review by hanlint — COMMENTED** [review]
Looks good -- can you run the same progress bar examples as here: https://github.com/mosaicml/composer/pull/1190

to check for edge cases?

**Comment by hanlint** [comment]
```
Epoch     0 eval/batch      100%|█████████████████████████| 4/4 [00:00<00:00,  5.12ba/s]                                                                                                                        
Epoch     0 train           100%|█████████████████████████| 3/3 [00:03<00:00,  1.32s/ba, loss/train=2.2866]                                                                                                     
Epoch     0 eval/ep         100%|█████████████████████████| 4/4 [00:00<00:00,  4.73ba/s]
```

This is a bit counter-intutive to me.. can we group all the `eval` bars together?

File: composer/loggers/progress_bar_logger.py [hunk 3]
```diff
@@ -229,62 +225,67 @@ def _build_pbar(self, state: State, is_train: bool, epoch_style: bool = False) -
...
 
             unit = TimeUnit.BATCH
             n = state.timestamp.epoch.value
-            if not is_train:
-                # eval results refer to model from previous epoch (n-1)
-                n = max(0, n - 1)
-            desc = f'Epoch {n:5d} {split:5s}'
+            if self.train_pbar is None and not is_train:
+                # epochwise eval results refer to model from previous epoch (n-1)
+                n -= 1
```
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/loggers/progress_bar_logger.py:249]: The `max` was needed before because mid-epoch evaluation progress bars followed this codepath. With the refactor, `self.train_pbar` would be not None if evaluating mid-epoch, causing this line to be skipped. 

Verified this works manually:

```yaml
train_dataset:
  &train_dataset
  mnist:
    use_synthetic: true
evaluators:
  eval1:
    label: eval/batch
    eval_interval: 1ba
    subset_num_batches: 2
    eval_dataset:
      <<: *train_dataset
model:
  mnist_classifier:
    num_classes: 10
max_duration: 1ep
train_batch_size: 2000
eval_batch_size: 2000
dataloader:
  pin_memory: true
  timeout: 0
  prefetch_factor: 2
  persistent_workers: true
  num_workers: 1
train_subset_num_batches: 2
```

Output:
```
Epoch   0, Batch   1: eval/batch      100%|█████████████████████████| 2/2 [00:00<00:00,  5.45ba/s]                                                                                                   
Epoch   0, Batch   2: eval/batch      100%|█████████████████████████| 2/2 [00:00<00:00,  5.22ba/s]                                                                                                   
Epoch   0:            train           100%|█████████████████████████| 2/2 [00:01<00:00,  1.38ba/s, loss/train=2.2977] 
```

**Comment by ravi-mosaicml** [comment]
> ```
> Epoch     0 eval/batch      100%|█████████████████████████| 4/4 [00:00<00:00,  5.12ba/s]                                                                                                                        
> Epoch     0 train           100%|█████████████████████████| 3/3 [00:03<00:00,  1.32s/ba, loss/train=2.2866]                                                                                                     
> Epoch     0 eval/ep         100%|█████████████████████████| 4/4 [00:00<00:00,  4.73ba/s]
> ```
> 
> This is a bit counter-intutive to me.. can we group all the `eval` bars together?

I agree it's a bit messy, but it happens since epoch-wise evaluation occurs after Event.EPOCH_END. The training progress bar closes, and then the epoch-wise evaluation runs. I updated the progress bar styling to include more timing information to make it clearer what is happening.

File: composer/loggers/progress_bar_logger.py [hunk 4]
```diff
@@ -31,22 +32,26 @@ def __init__(
...
         self.pbar = tqdm.auto.tqdm(
             total=total,
             position=position,
             bar_format=bar_format,
             file=file,
             dynamic_ncols=True,
+            # We set `leave=False` so TQDM does not jump around, but we emulate `leave=True` behavior when closing
+            # by printing a dummy newline and refreshing to force tqdm to print to a stale line
+            # But on k8s, we need `leave=True`, as it would otherwise overwrite the pbar in place
+            leave=not os.isatty(self.file.fileno()),
```
[the remark this answers is not in the document]
Comment (siriuslee) [reply composer/loggers/progress_bar_logger.py:52]: This doesn't quite work at the moment, though. Looks like the `dynamic_ncols` probably doesn't get set appropriately in TTY-enabled containers, so the progress bar shows up like:
```
100% 29/29 [00:00<00:00, 43.00ba/s, loss/train=0.2813]
100% 10/10 [00:00<00:00, 80.87ba/s, metrics/eval/Accuracy=0.9125]
```
I'm sure we could work around this, though.

**Review by siriuslee — APPROVED** [review]
LGTM. Really great work! Tested on MNIST in K8s and it works perfectly.

**Review by hanlint — APPROVED** [review]
LGTM, one UX question:

when `eval_interval = 1ba`, i think the user expects the timestamp to be ba1, ba2, ba3.. not the `batch_in_epoch` your current example emits:
```
eval/batch     Epoch   0, Batch   1:  100%|█████████████████████████| 2/2 [00:00<00:00,  4.71ba/s]                                                                                                   
eval/batch     Epoch   0, Batch   2:  100%|█████████████████████████| 2/2 [00:00<00:00,  5.45ba/s]                                                                                                   
train          Epoch   0:             100%|█████████████████████████| 2/2 [00:01<00:00,  1.22ba/s, loss/train=2.3006]                                       
```
The common use case here is `max_duration=1ep, eval_interval=1ba`, which would lead to repeated `Epoch 0` cluttering the progress bar.

could this be simplified (and also lead to a less wide pbar)?

File: composer/loggers/progress_bar_logger.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/loggers/progress_bar_logger.py:52]: Hmm, let's leave `tty: False` for now, since that works. Would like to get this fix out and then make it better :)

File: composer/loggers/progress_bar_logger.py [hunk 5]
```diff
@@ -161,30 +172,16 @@ def __init__(
...
-            self.eval_pbar = pbar
-
     @property
     def show_pbar(self) -> bool:
         return self._show_pbar and dist.get_local_rank() == 0
 
     def log_data(self, state: State, log_level: LogLevel, data: Dict[str, Any]) -> None:
         # log to progress bar
-        if self._current_pbar:
+        current_pbar = self.eval_pbar if self.eval_pbar is not None else self.train_pbar
```
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/loggers/progress_bar_logger.py:181]: Yes, the check on lines 182 / 184 is to update the metrics in the pbar.

current_pbar would be None if doing a separate pbar for each epoch, and something is being logged between when the closes the previous epoch and starts the new one.

File: composer/loggers/progress_bar_logger.py [hunk 6]
```diff
@@ -293,18 +300,33 @@ def eval_batch_end(self, state: State, logger: Logger) -> None:
...
-            self._close()
+        # Only close progress bars at epoch end if the duration is in epochs, since
+        # a new pbar will be created for each epoch
+        # If the duration is in other units, then one progress bar will be used for all of training.
+        assert state.max_duration is not None, 'max_duration should be set'
+        if self.train_pbar and state.max_duration.unit == TimeUnit.EPOCH:
+            self.train_pbar.close()
+            self.train_pbar = None
+
+    def close(self, state: State, logger: Logger) -> None:
```
[the remark this answers is not in the document]
Comment (ravi-mosaicml) [reply composer/loggers/progress_bar_logger.py:311]: Good catch, thanks. Usually the eval bar would be closed, but potentially not if there was a crash and eval_end did not run.
